File: POS_THEORIE/Semester2/Kap4/Praxis5/Kreis.c
/**
* C-Programm, gleichzustellen mit gleicher Funktionalitaet der Java-Klassen Koordinate, Kreis und Main
**/

#include <stdio.h>  // ermoeglicht Ausgabe usw.
#include <math.h>   // fuer mathematische Funktionen (unten: sqrt)

// Struktur mit 2 Integer; gleichzustellen mit Koordinate.java
struct koordinate {
    int x;
    int y;
};

// Struktur aus einer "Instanz" der Koordinatenstruktur und einem int Radius; gleichzustellen mit Kreis.java
struct kreis {
    koordinate coord;
    int radius;
};

// Parameter: Integer-Pointer
// mit Hilfe dieser werden die Datenfelder der main()-Methode manipuliert
void eingabe(int* x, int* y, int* z) {
    printf("\nPosition (Koordinate in der Mitte des Kreises):\n");
    // Eingabe NORMALERWEISE mit &
    // z.B scanf("%d", &x);
    // wird aber durch die Pointer-Parameter schon mitgegeben ???
    printf("X-Koordinate: ");
        scanf("%d", x);
    printf("Y-Koordinate: ");
        scanf("%d", y);
    printf("Radius eingeben: ");
        scanf("%d", z);
}

// Parameter: Kreisstruktur-Pointer und Integer-Pointer
// mit Hilfe dieser werden die Datenfelder der main()-Methode manipuliert
void setPosition(kreis* c, int* newX, int* newY, int* newRadius) {
    // coord ist ein Datenfeld der Kreisstruktur-Instanz c                      ->
    // x und y sind Datenfelder der Koordinatenstruktur-Instanz coord in c      .
    c->coord.x = *newX;
    c->coord.y = *newY;
    c->radius = *newRadius;
}
r
// Parameter: Kreisstruktur-Pointer
// Werte werden wie in setPosition() aufgerufen und dann ausgegeben
void output(kreis* c) {
    const double PI = 3.141592653589793;
    printf("\n**************************************");
    printf("\nMitte M(%d|%d)", c->coord.x, c->coord.y);
    printf("\nRadius r = %d", c->radius);
    printf("\nUmfang u = %lu \nFlaeche f = %f", (2 * PI * c->radius), (PI * sqrt(c->radius)));
    printf("\n**************************************\n");
}

// Parameter: Integer-Pointer
void menue(int* user) {
    printf("\n--------------------------------------");
    printf("\n1 - Werte des Kreises aendern");
    printf("\n2 - Informationen ausgeben");
    printf("\n3 - Programm Beenden");
    printf("\n--------------------------------------");
    printf("\nWelche Option? ");
    scanf("%d", user); // hier wieder kein & benoetigt !!!
}

// main()-Methode
int main() {

    int user = 0;
    int x = 0;
    int y = 0;
    int z = 0;
    kreis circle;

    // bei Methodenaufrufen wird immer die ADRESSE mitgeliefert
    eingabe(&x, &y, &z);
    setPosition(&circle, &x, &y, &z);

    do {
        menue(&user);
        switch(user) {
            case 1: eingabe(&x, &y, &z); setPosition(&circle, &x, &y, &z); break;
            case 2: output(&circle); break;
        }
        // printf("%d %d %d %d", user, x, y, z); // testing
    } while(user != 3);

    printf("\nByeBye!\n");

}

File: POS_THEORIE/Semester2/Kap4/Praxis5/Kreis.java
import java.util.Scanner;

/**
* Kreis-Klasse mit einem Datenfeld des Typs Koordinate und einem int Radius
**/
public class Kreis {

    private Koordinate mitte;   // durch die toString()-Methode und fehlenden Bedarf muessen keine Getter geschrieben werden
    private int radius;

    /**
    * Konstruktor der Kreis-Klasse
    * @param mitte Initialisierung des Datenfeldes der Klasse
    * @param radius Initialisierung des Datenfeldes der Klasse
    **/
    public Kreis(Koordinate mitte, int radius) {
        this.mitte = mitte;
        this.radius = radius;
    }

    /**
    * Methode zum Neu-Setzen der Datenfelder
    * @param newX neues x der Koordinate mitte
    * @param newY neues y der Koordinate mitte
    * @param radius neuer Wert fuer das radius Datenfeld dieser Klasse
    **/
    public void setPosition(int newX, int newY, int newRadius) {
        this.mitte.setCoordinate(newX, newY);
        this.radius = newRadius;
    }

    /**
    * ueberschreibt die toString()-Methode der Object-Klasse
    * benutzt die toString()-Methode der Koordinate-Klasse und fuegt die restlichen Infos der Kreis-Klasse hinzu
    * @return der fertige String
    **/
    @Override
    public String toString() {
        String output = "";
        output += "\n**************************************\n";
        output += this.mitte;
        output += "Radius r = " + this.radius + "\n";
        output += "Umfang u = " + (2 * Math.PI * this.radius) + "\nFlaeche f = " + (Math.PI * Math.pow(this.radius, 2));
        output += "\n**************************************";
        return output;
    }

}